Units Page,  Select Component Dropdowns, New Graphic Rate Dropdowns

Deelane opened the pull request:
# Description

### Units Page

Overhauled the units page as introduced in PR 766, including additional functionality, bugfixes, and updates. The page is now complete and is being used as the template for the creation of similar pages like the Meters and Conversions pages. 

Notable changes are as follows:
- Components converted to hooks and now uses hooks to access and alter Redux store and local state.
- Container files removed.
- Actions optimized, unnecessary actions removed.
- State/Reducers optimized, unnecessary state removed.
- Fixed bugs with reducers not properly updating store state.
- Updated UI visuals.

### Meters Page (TODO)

Updated to match units page functionality and interface using units page as template. This is partly done at this time in code included in this PR.

### Select Component Dropdowns

Added functionality and fixed bugs associated with selecting groups, meters, and units from the dropdowns in the ChartDataSelectComponent. The component should now work as intended.

Notable changes are as follows:
- Added functions for determining compatibility for groups and meters with other selected items.
- Fixed bugs where store state was not properly updated.
- Fixed bugs associated with unselecting a unit.
- Select items are now properly disabled if they are incompatible with other items.
- Added functionality allowing for switching between units without clearing the unit.
- Clearing the selected unit now clears all meters and groups, resetting the graph.

### New Graphic Rate Menu

Added a new rate menu allowing for changing of the time unit for line graphs.

### Issues Addressed

Partly Addresses #726
TODO Fixes #644

### Developer test data

Code was added to create units, meters and groups for testing the updated system. I directly adds via code rather than using CSV files (previous CSV used for this are now removed). The script testData now uses this code.

TODO Need some more added to deal with rate changes because kWh and kW are treated specially.

## Type of change

- [ ] Note merging this changes the node modules
- [ ] Note merging this changes the database configuration.
- [x] This change requires a documentation update

## Checklist


- [x] I have followed the [OED pull request](https://openenergydashboard.github.io/developer/pr.html) ideas
- [x] I have removed text in ( ) from the issue request

## Limitations

The TODOs must still be addressed.

Commit:
added files for admin unit page

andrew-cline committed:
Merge branch 'OpenEnergyDashboard:resourceGeneralization' into resourceGeneralization

Commit:
added front-end starter code for admin units page

Commit:
added necessary files for displaying units

Commit:
finished unit display page, added page and form for adding units

Commit:
Merge branch 'resourceGeneralization' of https://github.com/OpenEnergyDashboard/OED into resourceGeneralization

JustLe56 committed:
added ES and FR translation messages for unit page

Commit:
edited UnitsDetailContainer

Commit:
edited UnitsDetailContainer

sagarprasad63574 committed:
edited units details component, container, etc

sagarprasad63574 committed:
Merge pull request #1 from andrew-cline/sagar

edited units details component, container, etc

JustLe56 committed:
added header and footer to UnitDetailComponent

andrew-cline committed:
Merge pull request #2 from andrew-cline/resourceGenJustin

added header and footer to UnitDetailsComponent

sagarprasad63574 committed:
testing

sagarprasad63574 committed:
Merge branch 'resourceGeneralization' of https://github.com/andrew-cline/OED into sagar

sagarprasad63574 committed:
IDK

sagarprasad63574 committed:
Merge pull request #3 from andrew-cline/sagar

Sagar

sagarprasad63574 committed:
Added the button to identifier. Need to fix the logic

JustLe56 committed:
added SecInRate update; updates in client but does not save to database yet

Commit:
added button to unitRepresent column in UnitViewComponent

andrew-cline committed:
Merge pull request #4 from andrew-cline/sagar

Added the button to identifier. Need to fix the logic

andrew-cline committed:
Merge branch 'resourceGeneralization' into resourceGenJustin

andrew-cline committed:
Merge pull request #5 from andrew-cline/resourceGenJustin

Add SecInRate edit button

andrew-cline committed:
Revert "Add SecInRate edit button"

andrew-cline committed:
Merge pull request #6 from andrew-cline/revert-5-resourceGenJustin

Revert "Add SecInRate edit button"

andrew-cline committed:
Revert "Added the button to identifier. Need to fix the logic"

andrew-cline committed:
Merge pull request #7 from andrew-cline/revert-4-sagar

Revert "Added the button to identifier. Need to fix the logic"

Commit:
Merge branch 'resourceGeneralization' of https://github.com/andrew-cline/OED into Andrew

Hardik-Kharpude committed:
added a  button to the notes column in units page

andrew-cline committed:
Merge pull request #8 from andrew-cline/Andrew

Andrew

Commit:
merged resourceGenJustin with Andrew

andrew-cline committed:
Merge pull request #9 from andrew-cline/Andrew

merged resourceGenJustin with Andrew

Commit:
added UnitRepresent button

andrew-cline committed:
Merge pull request #10 from andrew-cline/Andrew

added UnitRepresent button

sagarprasad63574 committed:
Made changes to the backend

sagarprasad63574 committed:
Edited the backend and updated the code to work with unit represent and sec in rate

andrew-cline committed:
Merge branch 'resourceGeneralization' into HardikNoteButton

andrew-cline committed:
Merge pull request #12 from andrew-cline/HardikNoteButton

added a  button to the notes column in units page

Commit:
fixed somethings leftover from merge

Commit:
edited routes to include check for admin privledge

andrew-cline committed:
Merge pull request #13 from andrew-cline/Andrew

edited routes to include check for admin privledge

Commit:
added validation for secInRate

andrew-cline committed:
Merge pull request #14 from andrew-cline/Andrew

added validation for secInRate

JustLe56 committed:
fixed status message displaying on all columns

JustLe56 committed:
fixed checkbox; add submitNewUnit; changed unit id and identifer definition, added addUnit func to api

Commit:
fixed bug in unit create form

Commit:
fixed spelling in translation file

JustLe56 committed:
misc text fixes

JustLe56 committed:
added eng message for create unit container

JustLe56 committed:
fixes default selected dropdowns not triggering onchange

andrew-cline committed:
Merge branch 'OpenEnergyDashboard:resourceGeneralization' into resourceGeneralization

andrew-cline committed:
Merge branch 'resourceGeneralization' into sagar

andrew-cline committed:
Merge pull request #11 from andrew-cline/sagar

Made changes to the backend

Commit:
Merge branch 'justin_add_units' of https://github.com/andrew-cline/OED into resourceGeneralization

Commit:
Merge branch 'resourceGeneralization' of https://github.com/andrew-cline/OED into resourceGeneralization

Commit:
added formatted messages to createUnitComponent

Commit:
fixed preferredDisplay in UnitViewComponent

Commit:
Merge branch 'resourceGeneralization' of https://github.com/OpenEnergyDashboard/OED into resourceGeneralization

Commit:
removed duplicate unit files

Commit:
updated app.js to use units.js

Commit:
added jsonSchema import to units.js

JustLe56 committed:
changed add unit link to a modal display; need to add form content

JustLe56 committed:
adds modal create unit form; currently not working needs container

JustLe56 committed:
fixed imports

JustLe56 committed:
modal variation using hooks, does not actually modify database yet

JustLe56 committed:
create unit modal successfully submits to database; updated create unit form defaults

sagarprasad63574 committed:
UnitModelEdit

JustLe56 committed:
polished code formatting and added some comments

JustLe56 committed:
fixed unit state issue

Commit:
fixed unit state issue

JustLe56 committed:
Merge branch 'justin_createUnitModal' of https://github.com/andrew-cline/OED into clean_oed_resgen

JustLe56 committed:
Merge branch 'Unit_Dropdown' of https://github.com/andrew-cline/OED into createUnitModalCopy

sagarprasad63574 committed:
Merge branch 'resourceGeneralization' of https://github.com/andrew-cline/OED into UnitModelEdit

Commit:
added group dropdown logic and unit to group state

Commit:
edited submitNewUnit and addUnit route

sagarprasad63574 committed:
Working with unit model on the unit page and unit edit. Some problems are in order to update the unit in the database, need to click the save changes button twice(not sure where the problem is occurring)

JustLe56 committed:
lint changes

JustLe56 committed:
more linter changes

JustLe56 committed:
Merge branch 'resourceGeneralization' of https://github.com/OpenEnergyDashboard/OED into createUnitModal_UnitDropdown_merged

JustLe56 committed:
included edit note in render

JustLe56 committed:
added comments and lint changes

Hardik-Kharpude committed:
added group dropdown

Commit:
cleaned up group dropdown code

Commit:
Merge branch 'resourceGeneralization' of https://github.com/OpenEnergyDashboard/OED into group_dropdown

Commit:
removed extra lines in one function

sagarprasad63574 committed:
Bug Fixes and Editing the modal for the units page and edit unit

sagarprasad63574 committed:
Bug Fixes

JustLe56 committed:
Merge branch 'UnitModelEdit' of https://github.com/andrew-cline/OED into modalEdit_modalCreate

JustLe56 committed:
Merge branch 'UnitModelEdit' of https://github.com/andrew-cline/OED into modalEdit_modalCreate

huss committed:
Get groups mostly working with units

- Changed so when OED gets group details it gets the deep meters.
This happens on startup so that state is always there.
- Remove async from metersInGroup as should not have such a function in
mapStateToProps. This was causing a bug where the lack of an await meant
the groups were processed before the meters were know. Always having
deep meters with groups means you don't need to go to the DB.
- Changed sorting for menu display for groups and made unit be the same.
- Fixed bug in how compatible groups and units for sets determined.
- Added script (not yet fully ready) to add meters and readings for
testing graphing of data with units. Has CSV file with data.
- Added testing functions for Cik and Pik where do not need to pass
a conn.

JustLe56 committed:
added linter fixes

JustLe56 committed:
added documentation and cleaned up comments

sagarprasad63574 committed:
Bug fixes

sagarprasad63574 committed:
Merge branch 'modalEdit_modalCreate' of https://github.com/andrew-cline/OED into modalEdit_modalCreate

huss committed:
fix check issue

JustLe56 committed:
linter fixes

Commit:
added suggested changes

Commit:
Merge branch 'group_dropdown' of https://github.com/andrew-cline/OED into group_dropdown

huss committed:
See PR for what was done

huss committed:
remove unique constaint

This was causing an issue when units were created/edited so it will
now be enforced in software. See the design document for more details.
The code still needs to add an update of cik/pik on unit changes.

huss committed:
remove left over trailing comma

JustLe56 committed:
updated translations

JustLe56 committed:
removed unit.create.new.unit

Commit:
updated create and edit unit forms to use enums

Commit:
removed unused import

Commit:
made changes based on comments on pull request

JustLe56 committed:
changed createUnitModalComponent form

JustLe56 committed:
Merge branch 'modalEdit_modalCreate' of https://github.com/andrew-cline/OED into modalEdit_modalCreate

huss committed:
updates (see details below)

- translated the labels of items
- translated enum menu items per upcoming changes in other parts
of the code base.
- translated Yes/No menu dropdown
- translated some buttons
- reorder edit page so items are same as create page and model

huss committed:
Merge branch 'resourceGeneralization' of https://github.com/OpenEnergyDashboard/OED into pr/andrew-cline/766-1

This should update the PR so it is now in sync with the
resourceGeneralization branch that is also in sync with development
branch. It should help those working on the modal issue.

wootent committed:
commit with updated changes from pr#766 and re implimented changes

wootent committed:
updated createUnitModal

wootent committed:
fixed SQL to update/edit units - name of unit was not able to change, it would update in state but not write to DB so upon a refresh the name would change back to previous.

wootent committed:
cleaned up create unit modal container & componet

wootent committed:
added function to refresh page after unit is created

wootent committed:
ProtoTyped color code display on Unit cards

wootent committed:
small change to last edit

Deelane committed:
Pre-commit commit. Made updates to the unit page: Removed the need for containers (left container files for now). Updated all components to functional hooks, updated Redux state management to new style (useSelector, useDispatch). Cleaned up code for better readability. Added action for adding new units.

Deelane committed:
Fixed EditUnitModalComponentProps misnamed in file

Deelane committed:
Added notifications for success and failure of adding/editing units

Deelane committed:
Updated initial units fetch, added new action

wootent committed:
changed close button to dicard changes and CSS update

wootent committed:
update edit unit modal - save all btn disabled when no values for namme, identifier

wootent committed:
update to edit modal

Deelane committed:
Fixed bugs causing store not to properly update, fixed units page not updating due to inproper store state updating, split confirmEditUnits action into multiple actions that can now account for failure to edit unit

Deelane committed:
removed old comment

Deelane committed:
Merge branch 'twoot_766_fork' into pr-766-test-changes

Deelane committed:
Merge pull request #5 from Deelane/pr-766-test-changes

merge Deelane 766 -> wootent 766

Deelane committed:
Merged wootent and deelane pr766 changes

Deelane committed:
fixed sorting to account for lowercase

wootent committed:
fix perferred display

wootent committed:
Merge branch 'pr-766-wootent-deelane-merge' of https://github.com/Deelane/OED into pr-766-wootent-deelane-merge

Deelane committed:
added english mappings for success/failure to edit and create units

Deelane committed:
Removed unnecessary state variable edited units and associated actions/reducers. Also removed bulk submit edited units action. The store should never know about edits that other people are making, and the previous functionality also introduced a bug where two unit edits submitted at the same time was very indeterminate, now it is just less indeterminate

Deelane committed:
Merge branch 'pr/andrew-cline/765' into units-page-group-dropdown-merge

huss committed:
get unit translation into all languages

huss committed:
help link fixed for units

huss committed:
add units to test data

- The testData command now loads data that has units.
- There are new simple test cases for developers.
- testData was converted from a script to code to handle units but
also makes it faster.
- A seeming mistake where the automated testing generation did not do
awaits was fixed.

Deelane committed:
Merge branch 'pr-766-wootent-deelane-merge' into units-page-group-dropdown-merge

Deelane committed:
unit identifier now defaults to unit name if left blank when creating a new unit

Deelane committed:
made html component fields per line, removed unused code

huss committed:
update units+

- Changed so bar does not return readings for raw in DB.
- Changed so can skip input of unit identifier and defaults to name.
Changed code to inforce this in Unit model.
- Create/use enum for true/false menus. Show as yes/no. Includes
translations.
- Changed Boolean type for CSV to yes/no so same.
- Sort by name not identifier for unit since that is highlighted in each
card.
- Remove admin check on unit view since route stops access by others.
- Changed values for tempurature test meter so should be straight
once graph raw.
- Comment out ports in Docker file.
- Removed name key on Input since not used at this point.
- Put each Input key on own line.
- Other small edits.

huss committed:
merge upstream

wootent committed:
reverse the order on the edit and create modal popups so the identifier comes before the name

wootent committed:
HTML format changes so edit & create modal appears as close as possiable visually, not done need assistance on edit page, identifier element does not render correctly

wootent committed:
put the identifier in the blue title of each card on view page and put the name in the list of unit values

wootent committed:
preferred display on two lines if false corrected

huss committed:
revert so sort units by identifier again

AustinStewart95 committed:
Added help icon to Edit unit modal.
Needs to be updated with correct text.

AustinStewart95 committed:
Merge branch 'units-page-group-dropdown-merge' of https://github.com/Deelane/OED into units-page-group-dropdown-merge

jbonbright committed:
Added background color to displayable in unit card

jbonbright committed:
Merge branch 'units-page-group-dropdown-merge' of https://github.com/Deelane/OED into units-page-group-dropdown-merge

Commit:
Reduce CreateUnit handlers to update many inputs

nanoarzqta committed:
Updated edit unit rate value to stop scroll or down arrow ability, cannot pass below 1 when scrolling or using down arrow key

jebonbright5 committed:
Merge branch 'units-page-group-dropdown-merge' of https://github.com/Deelane/OED into units-page-group-dropdown-merge

nanoarzqta committed:
Merge branch 'units-page-group-dropdown-merge' of https://github.com/Deelane/OED into units-page-group-dropdown-merge

Commit:
Reduce EditUnit handlers to update many inputs

Commit:
Reduce EditUnit handlers to update many inputs

huss committed:
sec in rate > 0

- Make seconds in rate > 0 on create input if use arrows to mirror edit.
Also make it a constraint in DB so can never happen.
- Remove edit check for
oninput="validity.valid||(value='');"
on edit since it was not working for negative values as hoped.
- Remove unique on unit DB table create on migration to mirror
other place.

AustinStewart95 committed:
Added help text icon to create unit modal.

wootent committed:
Meter Modal devlopment

wootent committed:
fixed meterview - added translate

wootent committed:
fixed css for meterview - added class for card-container

wootent committed:
fixed formmated messages

wootent committed:
UI updates

Deelane committed:
Changes to ChartDataSelectComponent: units/meters/groups selects now disabled if incompatible, non-admin selects now properly hidden, groups/meters without default graphic unit now disabled when no unit is selected, fixed bug where selecting unit first did not allow graphing, added buttons to clear selects

huss committed:
add test groups

wootent committed:
updates

huss committed:
update unit testData

- Allow meters & groups to control displayable value.
- Allow meters to not have a unit (-99).
- Added some meters/groups for new options/testing.

wootent committed:
Merge branch 'units-page-group-dropdown-merge' of https://github.com/Deelane/OED into units-page-group-dropdown-merge

Deelane committed:
Removed buttons, extraneous code and comments

huss committed:
fix missing units on meter in testData

huss committed:
Merge branch 'units-page-group-dropdown-merge' of https://github.com/Deelane/OED into units-page-group-dropdown-merge

Deelane committed:
Added first version of new rate menu allowing for the selection of different time rates for line graphs of specific unit types

huss committed:
update rate menu code

- Put back special label for kWh/hour by fixing check.
- Remove TODO on raw/flow since flow correctly handled above.
- Fix hover label so matches new rate value plotted.
- Add translations for hour, minute, second in French/Spanish.

huss committed:
Delete settings.json

We don't include these types of files to the repo.

huss commented on src/client/app/components/ChartDataSelectComponent.tsx:
```diff
@@ -279,49 +287,39 @@ export default function ChartDataSelectComponent() {
  * @return {SelectOption[]} an array of SelectOption
  */
 export function getUnitCompatibilityForDropdown(state: State) {
```
If this is only used in this file do we need to export?

Deelane replied:
Do not need to export, export removed.

huss commented on src/client/app/components/ChartDataSelectComponent.tsx:
```diff
@@ -332,50 +330,237 @@ export function getUnitCompatibilityForDropdown(state: State) {
+	}
+	// User is not an admin, do not allow for admin units to be seen
+	else {
+		visibleUnitsOrSuffixes = _.filter(state.units.units, (o: UnitData) => {
```
I think it might be easier to check for DisplayableType.all and that would also be safer if another type is added.

Deelane replied:
Change made.

huss commented on src/client/app/components/RouteComponent.tsx:
```diff
@@ -243,14 +244,16 @@ export default class RouteComponent extends React.Component<RouteProps> {
 								<Route path='/users/new' render={() => this.requireAuth(<CreateUserContainer/>)}/>
 								<Route path='/users' render={() => this.requireAuth(<UsersDetailContainer fetchUsers={() => []}/>)}/>
+								<Route path='/units'render={() => this.requireAuth(<UnitsDetailComponent />)}/>
+								{/* <Route path="/addUnit" render={() => this.requireAuth(<CreateUnitContainer />)}/> */}
```
Can this comment go?

Deelane replied:
Comment gone.

huss commented on src/client/app/components/unit/CreateUnitModalComponent.tsx:
```diff
@@ -0,0 +1,243 @@
+											name='suffix'
+											type='text'
+											onChange={e => handleStringChange(e)}
+											required value={state.suffix} />
```
I'm uncertain what this required is for, esp. since you can save without entering any value.

Deelane replied:
I agree, it does nothing and was a remnant of the previous code. Removed all instances of required.

huss commented on src/client/app/components/unit/EditUnitModalComponent.tsx:
```diff
@@ -0,0 +1,270 @@
+import { removeUnsavedChanges } from '../../actions/unsavedWarning';
+import TooltipMarkerComponent from '../TooltipMarkerComponent';
+import TooltipHelpContainer from '../../containers/TooltipHelpContainer';
+// I realize that * is already imported from react
```
Can we get rid of the "I" and maybe give a reason for needing the second import?

Deelane replied:
Removed the "I". The reason is purely visual for code readability. I can remove it if need be.

huss commented on src/client/app/components/unit/UnitViewComponent.tsx:
```diff
@@ -0,0 +1,72 @@
+import { FormattedMessage } from 'react-intl';
+import EditUnitModalComponent from './EditUnitModalComponent';
+import '../../styles/unit-card-page.css';
+// I realize that * is already imported from react
```
Same "I" comment as elsewhere.

Deelane replied:
Removed.

huss commented on src/client/app/components/unit/UnitViewComponent.tsx:
```diff
@@ -0,0 +1,72 @@
+			</div>
+			<div className={props.unit.preferredDisplay.toString()}>
+				<b><FormattedMessage id="unit.preferred.display" /></b> {translate(`TrueFalseType.${props.unit.preferredDisplay.toString()}`)}
+				{/* <b><FormattedMessage id="unit.preferred.display" /></b> {props.unit.preferredDisplay.toString()} */}
```
Comment needed?

Deelane replied:
Comment not needed, removed.

huss commented on src/client/app/containers/ChartDataSelectContainer.ts:
```diff
@@ -0,0 +1,346 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
```
This file was brought back but I cannot see any use and I would think it is gone given the new Component with hooks. Is it needed? It also has overlapping functions with those in the ccomponent.

Deelane replied:
File not needed, deleted it.

huss commented on src/client/app/types/redux/actions.ts:
```diff
@@ -114,8 +108,25 @@ export enum ActionType {
 	RequestUnitsDetails = 'REQUEST_UNITS_DETAILS',
-	ReceiveUnitsDetails = 'RECEIVE_UNITS_DETAILS'
+	ChangeDisplayedUnits = 'CHANGE_DISPLAYED_UNITS',
+	EditUnitDetails = 'EDIT_UNIT_DETAILS',
```
Is this used? I could not find it in a search.

Deelane replied:
Not used, removed.

huss commented on src/client/app/types/redux/actions.ts:
```diff
@@ -114,8 +108,25 @@ export enum ActionType {
+	EditUnitDetails = 'EDIT_UNIT_DETAILS',
+	SubmitEditedUnit = 'SUBMIT_EDITED_UNIT',
+	ConfirmEditedUnit = 'CONFIRM_EDITED_UNIT',
+	DeleteEditedUnit = 'DELETE_EDITED_UNIT',
```
Is this used? I could not find it in a search.

Deelane replied:
Not used, removed.

huss commented on src/client/app/types/redux/actions.ts:
```diff
@@ -114,8 +108,25 @@ export enum ActionType {
+	ReceiveMetersDetails = 'RECEIVE_METERS_DETAILS',
+	RequestMetersDetails = 'REQUEST_METERS_DETAILS',
+	ChangeDisplayedMeters = 'CHANGE_DISPLAYED_METERS',
+	EditMeterDetails = 'EDIT_METER_DETAILS',
```
Is this used? I could not find it in a search.

huss commented on src/client/app/types/redux/actions.ts:
```diff
@@ -114,8 +108,25 @@ export enum ActionType {
+	EditMeterDetails = 'EDIT_METER_DETAILS',
+	SubmitEditedMeter = 'SUBMIT_EDITED_METER',
+	ConfirmEditedMeter = 'CONFIRM_EDITED_METER',
+	DeleteEditedMeter = 'DELETE_EDITED_METER',
```
Is this used? I could not find it in a search.

wootent replied:
DeleteEditedMeter is the only action that is not currently in use, but I think its for future?

huss commented on src/client/app/utils/api/unitsApi.ts:
```diff
@@ -5,15 +5,35 @@
 		this.backend = backend;
 	}
 
+	public async details(): Promise<NamedIDItem[]> {
```
This seems very similar to getUnitsDetails. A search for unitsApi found a use of getUnitsDetails but not this one. Is it needed and what is it for?

Deelane replied:
Does not seem to be used. Removed.

huss commented on src/server/routes/units.js:
```diff
@@ -6,26 +6,191 @@ const express = require('express');
+		type: 'object',
+		required: ['name', 'identifier', 'unitRepresent', 'typeOfUnit', 'displayable', 'preferredDisplay'],
+		properties: {
+			id: {
```
id is not used as it must be set to undefined when adding to DB. I tried commenting this out and it does not matter. I know the value is sent because we send a unit) but would it be okay to remove it and put in a comment about why?

huss commented on src/server/routes/units.js:
```diff
@@ -6,26 +6,191 @@ const express = require('express');
+				minLength: 1,
+				enum: Object.values(Unit.unitType)
+			},
+			unitIndex: {
```
This is in the same category as id. The fact it may not be needed is seen from fact that number is mispelled.

Deelane replied:
Removed with no issues.

huss committed:
update testing & more

- Update how label kWh & kW and don't allow rate change on these.
- Fix so hoverText on line graphic shows scaling value.
- Added lots of missing translations outside English and removed a few
that no longer used.
- Put in unit create/edit tooltip help & links
- Remove test script and CSV files for unit tests as now included
in TesData
- Remove debug check that only one meter/group when selectedUnit is -99
- Some small changes to comment wording.
- Lots of small formatting changes.

huss commented:
I just pushed changes to the PR. See commit message/code changes for details.

I also note that the check is failing. All GitHub Action checks started failing recently so this is not an issue with this PR. I did run them locally and they all passed so I think the PR is okay.

My review should be posted soon.

huss reviewed (changes requested):
Overall this works as expected. I did not look at the meter code yet since it was not done. I have left some comments to address. Thanks for the good work.

Deelane committed:
Removed unused code, addressed comments

Deelane committed:
Removed more unused code, addressed comments

huss commented:
@wootent There are two comments on this PR that seem to involve code you added. They ask whether actions are needed since they don't seem to be used at this time. They are: EDIT_METER_DETAILS and DELETE_EDITED_METER. However, they might be used in upcoming work or I missed something. Could you let us know? Thanks.

huss commented:
The update by @Deelane covers everything involving them. The PR seems fine for units and the dropdown menus. It is on hold since it also has partly completed meter page work.

huss committed:
update test data & more

- Add Water quantity and flow units/conversions/meters to allow testing
of rate menu
- Allow rate for units for test data
- Always try to insert standard units in case missing for test data
- Fix up Postgres commands to remove all test data
- Add TODO about unused action

huss commented:
I just pushed changes for test data for rate meters and a comment on the extra action.

Deelane committed:
minor update to ensure submitting array is overwritten properly

Deelane committed:
small function call optimization

huss commented:
The meter work is in PR #786 so merging this so others can easily incorporate.
